app/models/user_manager.rb

```diff
@@ -0,0 +1,157 @@
+# Allows admins to manage user status and next of kin
+# via the admin users interface
+class UserManager
+  attr_reader :admin,
+              :user,
+              :kin_name,
+              :kin_email,
+              :admin_note,
+              :admin_action,
+              :suspension_length,
+              :errors,
+              :successes
+
+  def initialize(admin, params)
+    @admin = admin
+    @user  = User.find_by(login: params[:user_login])
+    @kin_name           = params[:next_of_kin_name]
+    @kin_email          = params[:next_of_kin_email]
+    @admin_note         = params[:admin_note]
+    @admin_action       = params[:admin_action]
+    @suspension_length  = params[:suspend_days]
+    @errors = []
+    @successes = []
+  end
+
+  def save
+    validate_user_and_admin &&
+      validate_admin_note &&
+      validate_suspension &&
+      validate_next_of_kin &&
+      save_next_of_kin &&
+      save_admin_action
+  end
+
+  def success_message
+    successes.join(" ")
+  end
+
+  def error_message
+    errors.join(" ")
+  end
+
+  private
+
+  def validate_user_and_admin
+    if user && admin
+      true
+    else
+      errors << "Must have a valid user and admin account to proceed."
+      false
+    end
+  end
+
+  def validate_admin_note
+    return true if admin_note.present? || admin_action.blank?
+    if admin_action == "spamban"
+      @admin_note = "Banned for spam"
+    elsif admin_action.present?
+      errors << "You must include notes in order to perform this action."
+      false
+    end
+  end
+
+  def validate_suspension
+    if admin_action == "suspend" && suspension_length.blank?
+      errors << "Please enter the number of days for which the user should be suspended."
+      false
+    else
+      true
+    end
+  end
+
+  # Basically, we either want valid data for both name and email
+  # or we want them both to be blank, which means do nothing
+  # or delete anything that already exists
+  def validate_next_of_kin
+    error = nil
+    if kin_name.present?
+      if kin_email.blank?
+        error = "Fannish next of kin email is missing."
+      elsif !User.where(login: kin_name).exists?
+        error = "Fannish next of kin user is invalid."
+      end
+    elsif kin_email.present?
+      error = "Fannish next of kin user is missing."
+    end
+    if error
+      errors << error
+      false
+    else
+      true
+    end
+  end
+
+  def save_next_of_kin
+    return true if user.fannish_next_of_kin.nil? && kin_name.blank?
+    if FannishNextOfKin.update_for_user(user, kin_name, kin_email)
+      successes << "Fannish next of kin was updated."
+    else
+      errors << "Fannish next of kin failed to update."
+      false
+    end
+  end
+
+  def save_admin_action
+    return true if admin_action.blank?
+    if admin_action == 'spamban'
+      ban_user
+    else
+      send("#{admin_action}_user")
```

Comment:
Do we do anything to restrict the possible values of `admin_action` so we don't risk people making up their own values by editing the form? (I don't think there's much chance of that, since this is an admin thing, but...)

Reply:
That works for me! When you have a minute, could you answer the other two questions on this PR? If they're okay, then it's ready to merge.

- https://github.com/otwcode/otwarchive/pull/2986#discussion_r134664590
- https://github.com/otwcode/otwarchive/pull/2986#discussion_r134665631

Reply:
Updated with changes